### PR TITLE
injector: Adding unit tests for the injector package

### PR DESCRIPTION
**Pull request by draychev**
This PR adds more unit test coverage for the `pkg/injector`

This PR also includes a rewrite of the `updateAnnotation()` function - to simplify and correct.
Previous iteration of this function returned `add` `JSONPatchOperation` with `Value` being sometimes `string` and other times ` map[string]string`. The most recent iteration always creates `add` `JSONPatchOperation` with `Value` of `map[string]string`.

### Before
```sh
$ go test -cover ./pkg/injector/...
ok      github.com/openservicemesh/osm/pkg/injector     0.058s  coverage: 27.6% of statements
```

### After
```sh
$ go test -cover ./pkg/injector/...
ok      github.com/openservicemesh/osm/pkg/injector     6.821s  coverage: 75.7% of statements
```

---

Please mark with X for applicable areas.

- New Functionality      [ ]
- Documentation          [ ]
- Install                [ ]
- Control Plane          [ ]
- CLI Tool               [ ]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests / CI System      [X]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution? `no`

### CODE REVIEW

**Comment by shashankram** [comment]
> This PR also includes a rewrite of the `updateAnnotation()` function - to simplify and correct.
> Previous iteration of this function returned `add` `JSONPatchOperation` with `Value` being sometimes `string` and other times ` map[string]string`. The most recent iteration always creates `add` `JSONPatchOperation` with `Value` of `map[string]string`.

This isn't incorrect but rather as expected. The way the function works prior to this change is that if `target` is nil, then no annotations exist, and hence the patch value is a map, ie. a new annotation is added. If the annotation exists, then the patch no longer needs to create a new map, and only needs to update the existing object with the `add` map.

File: pkg/injector/patch.go [hunk 1]
```diff
@@ -139,41 +135,67 @@ func addContainer(target, add []corev1.Container, basePath string) (patch []JSON
...
+type operation string
+
+func (o operation) String() string {
+	return string(o)
+}
+
+const (
+	addOperation     operation = "add"
+	replaceOperation operation = "replace"
+)
```
Comment (shashankram) [inline pkg/injector/patch.go:155]: I find this unnecessarily complicated, the patch operations are always strings so there doesn't seem to be a good reason why a new type needs to be defined for something as trivial. Please change this to `const strings`

File: pkg/injector/patch.go [hunk 2]
```diff
@@ -139,41 +135,67 @@ func addContainer(target, add []corev1.Container, basePath string) (patch []JSON
...
+
+			replaceOperation: func() interface{} {
+				return value
+			},
+		}[patchOperation]()
+
+		patches = append(patches, JSONPatchOperation{
+			Op:    patchOperation.String(),
+			Path:  patchPath,
+			Value: patchValue,
```
Comment (shashankram) [inline pkg/injector/patch.go:194]: I still don't understand what is wrong with the existing implementation, do you mind adding more details?

Check https://github.com/openservicemesh/osm/pull/1772#issuecomment-702252547

File: pkg/injector/webhook.go [hunk 3]
```diff
@@ -204,7 +211,9 @@ func (wh *webhook) mutate(req *v1beta1.AdmissionRequest) *v1beta1.AdmissionRespo
 
 	// Create the patches for the spec
 	// We use req.Namespace because pod.Namespace is "" at this point
-	patchBytes, err := wh.createPatch(&pod, req.Namespace)
+	// This string uniquely identifies the pod. Ideally this would be the pod.UID, but this is not available at this point.
+	proxyUUID := uuid.New().String()
+	patchBytes, err := wh.createPatch(&pod, req.Namespace, proxyUUID)
```
Comment (shashankram) [inline pkg/injector/webhook.go:216]: I'd prefer to revert this change since it seems unnecessary.

**Review by shashankram — CHANGES_REQUESTED** [review]
Thanks for adding the tests, some comments regarding the existing functionality.

File: pkg/injector/patch.go (hunk 2 above)
Comment (draychev) [reply pkg/injector/patch.go]: I noticed this function was non deterministic.  Decided to rewrite it - but it was not correct.
To add determinism I reverted to the correct way you had implemented previously @shashankram, but added an iteration over sorted map keys.

File: pkg/injector/webhook.go (hunk 3 above)
Comment (draychev) [reply pkg/injector/webhook.go:215]: @shashankram I noticed that `createPatch()` was creating the proxyUUID.  I think that `createPatch()` should not know what proxy UUID is or how it is generated. So I decided to pull it out of `createPatch()`.
This makes testing `createPatch()` a bit easier too.

File: pkg/injector/patch.go [hunk 4]
```diff
@@ -138,41 +139,68 @@ func addContainer(target, add []corev1.Container, basePath string) (patch []JSON
...
-			op := "add"
-			if target[key] != "" {
-				op = "replace"
-			}
-			patch = append(patch, JSONPatchOperation{
-				Op:    op,
-				Path:  path.Join(basePath, escapeJSONPointerValue(key)),
-				Value: value,
-			})
+func createTarget(key, path, value string) JSONPatchOperation {
```
Comment (shashankram) [inline pkg/injector/patch.go:150]: What target means here is ambiguous.

File: pkg/injector/patch.go [hunk 5]
```diff
@@ -138,41 +139,68 @@ func addContainer(target, add []corev1.Container, basePath string) (patch []JSON
...
+	return keys
+}
+
+func updateAnnotation(target, add map[string]string, basePath string) []JSONPatchOperation {
+	var patches []JSONPatchOperation
+
+	// If the target does not exist we need to create it
+	noTarget := len(target) == 0
+
+	// We iterate over the sorted keys in order to have a deterministic output from this function.
```
Comment (shashankram) [inline pkg/injector/patch.go:185]: Why does the output from this function need to be deterministic? If it is for the purpose of testing, I would add this to the comment.

File: pkg/injector/patch.go [hunk 6]
```diff
@@ -138,41 +139,68 @@ func addContainer(target, add []corev1.Container, basePath string) (patch []JSON
...
-				Op:    op,
-				Path:  path.Join(basePath, escapeJSONPointerValue(key)),
-				Value: value,
-			})
+func createTarget(key, path, value string) JSONPatchOperation {
+	return JSONPatchOperation{
+		Op:   addOperation,
+		Path: path,
+		Value: map[string]string{
+			key: value,
```
Comment (shashankram) [inline pkg/injector/patch.go:155]: Function is very specific to patching map types in API objects, so I would want to include the type (map) in the function name.

**Review by shashankram — COMMENTED** [review]
Looks correct overall, few comments and some linting issues to resolve.

File: pkg/injector/patch.go [hunk 7]
```diff
@@ -138,41 +139,71 @@ func addContainer(target, add []corev1.Container, basePath string) (patch []JSON
...
+// createMapPatchOperation is used specifically for the very first 'add' operation
+// for a non-existent 'path'.
+func createMapPatchOperation(key, path, value string) JSONPatchOperation {
+	return JSONPatchOperation{
+		Op:   addOperation,
+		Path: path,
+		Value: map[string]string{
+			key: value,
+		},
+	}
```
Comment (shashankram) [inline pkg/injector/patch.go:159]: I would very much like this to be moved back to how it was, to be consistent with the other patch functions. I see no reason for this to be a separate function.

**Review by shashankram — APPROVED** [review]

**Review by snehachhabria — APPROVED** [review]

File: pkg/injector/patch_test.go [hunk 8]
```diff
@@ -1,36 +1,239 @@
...
+			_, err := client.CoreV1().Namespaces().Create(context.TODO(), testNamespace, metav1.CreateOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			cache := make(map[certificate.CommonName]certificate.Certificater)
+
+			wh := &webhook{
+				kubeClient:     client,
+				kubeController: mockNsController,
+				certManager:    tresor.NewFakeCertManager(&cache, mockConfigurator),
+				meshCatalog:    catalog.NewFakeMeshCatalog(client),
```
Comment (eduser25) [inline pkg/injector/patch_test.go:190]: You know we have a mock for catalgo to right? >:)
